### PR TITLE
Remove metrics guards and add documentation

### PR DESCRIPTION
**Pull request by Algunenano**
PR to remove the need to use #ifdef's all the time and adding basic documentation on how to add new metrics.

### CODE REVIEW

File: include/mapnik/image_impl.hpp [hunk 1]
```diff
@@ -99,10 +99,8 @@ image<T>::image(image<T> const& rhs)
       offset_(rhs.offset_),
       scaling_(rhs.scaling_),
       premultiplied_alpha_(rhs.premultiplied_alpha_),
-      painted_(rhs.painted_)
-#ifdef MAPNIK_METRICS
-     ,metrics_(rhs.metrics_)
-#endif
+      painted_(rhs.painted_),
+      metrics_(rhs.metrics_)
```
Comment (rafatower) [inline include/mapnik/image_impl.hpp:103]: shouldn't this be guarded?

File: include/mapnik/image_impl.hpp [hunk 2]
```diff
@@ -112,10 +110,8 @@ image<T>::image(image<T> && rhs) noexcept
       offset_(rhs.offset_),
       scaling_(rhs.scaling_),
       premultiplied_alpha_(rhs.premultiplied_alpha_),
-      painted_(rhs.painted_)
-#ifdef MAPNIK_METRICS
-     ,metrics_(std::move(rhs.metrics_))
-#endif
+      painted_(rhs.painted_),
+      metrics_(std::move(rhs.metrics_))
```
Comment (rafatower) [inline include/mapnik/image_impl.hpp:114]: same here? and there are other cases below in this file

File: docs/metrics.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+#### Time
+
+Returns  an unique_ptr to class that will automatically add the metric once it's out of scope. This call will increase the "Time (us)" metric with the time passed and increase the "Calls" metric. In order to remove warnings when the compilation is done without `ENABLE_METRICS` please capture the metric with `METRIC_UNUSED auto`.
+    For example:
+
+```c++
+    for (int i = 0; i < 10; i++)
+    {
+        METRIC_UNUSED auto t = metrics_.measure_time("Mapnik.Setup");
```
Comment (rafatower) [inline docs/metrics.md:54]: It reads a little bit weird. how about `IF_METRICS_ENABLED`?

File: include/mapnik/metrics.hpp [hunk 4]
```diff
@@ -48,6 +46,35 @@ enum measurement_t : int_fast8_t
     TOTAL_ENUM_SIZE
 };
 
+#if __has_cpp_attribute(maybe_unused)
+#define METRIC_UNUSED [[maybe_unused]]
+#elif __has_cpp_attribute(gnu::unused)
+#define METRIC_UNUSED [[gnu::unused]]
+#else
+#define METRIC_UNUSED
+#endif
```
Comment (rafatower) [inline include/mapnik/metrics.hpp:55]: I don't quite understand how this macro works. Could you please explain it to me? (or point to some documentation).

At the same time, I'm a bit wary of breaking mac builds (which I cannot test in travis right now).

**Review by rafatower — COMMENTED** [review]
> metrics without guards

do you think it is worth it? no strong opinion on my side.

There are already some tests for the metrics themselves, aren't there?

File: include/mapnik/metrics.hpp (hunk 4 above)
Comment (Algunenano) [reply include/mapnik/metrics.hpp:55]: The whole idea with the PR is to not need guards anymore by setting two options with the metrics: the full functionality if the option is enabled or a dummy one simple enough to be (hopefully) removed automatically by the compiler if not.

This macro is  used to remove warnings of unused variable declarations if the code is compiled without metrics support. It uses cpp++17 attribute if avaible, if not it tries to use a gcc attribute (also available in clang) and if not it just doesn't do anything and the warning will be shown.

**Comment by Algunenano** [comment]
> do you think it is worth it? no strong opinion on my side.
Having the ifdef's everywhere is a PITA as a single line for a metric becomes 3.

File: include/mapnik/image_impl.hpp (hunk 1 above)
Comment (Algunenano) [reply include/mapnik/image_impl.hpp:103]: Not anymore, the idea is that if compiled without metrics the declaration is so simple that the compiler should be able to basically remove it or reduce its impact to 0.

**Review by rafatower — APPROVED** [review]

**Comment by rafatower** [comment]
@mtehver can you please take a look at this code?

**Comment by mtehver** [comment]
I took a close look at the code. Looks good to me.

I am not entirely sure that omitting the #ifdef MAPNIK_METRICS makes the code more readable, I personally would prefer #ifdef/#endif, but that is a matter of personal taste.  Also, pretty sure that performance wise the `measure_time` calls will be optimized fully away when using -O flags. So in that sense nothing wrong with the approach.

**Review by mtehver — COMMENTED** [review]
Looks good, no issues with the code or documentation.

**Pull request merged**